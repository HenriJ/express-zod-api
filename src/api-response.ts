import { z } from "zod";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
};

export interface ApiResponse<S extends z.ZodTypeAny> {
  schema: S;
  /**
   * @default 200 for positive response
   * @default 400 for negative response
   * */
  statusCodes?: number | [number, ...number[]];
  /** @default "application/json" */
  mimeTypes?: string | [string, ...string[]];
  /** @deprecated use statusCodes instead */
  statusCode?: never;
  /** @deprecated use mimeTypes instead */
  mimeType?: never;
}

export interface NormalizedResponse {
  schema: z.ZodTypeAny;
  statusCodes: [number, ...number[]];
  mimeTypes: [string, ...string[]];
}

export type AnyResponseDefinition =
  | z.ZodTypeAny // plain schema, default status codes applied
  | ApiResponse<z.ZodTypeAny> // single response definition, status code(s) customizable
  | ApiResponse<z.ZodTypeAny>[]; // Feature #1431: different responses for different status codes

export const normalizeApiResponse = (
  subject: AnyResponseDefinition,
  fallback: Omit<NormalizedResponse, "schema">,
): NormalizedResponse[] => {
  if (subject instanceof z.ZodType) {
    return [{ ...fallback, schema: subject }];
  }
  return (Array.isArray(subject) ? subject : [subject]).map(
    ({ schema, statusCodes, mimeTypes }) => ({
      schema,
      statusCodes:
        typeof statusCodes === "number"
          ? [statusCodes]
          : statusCodes || fallback.statusCodes,
      mimeTypes:
        typeof mimeTypes === "string"
          ? [mimeTypes]
          : mimeTypes || fallback.mimeTypes,
    }),
  );
};
