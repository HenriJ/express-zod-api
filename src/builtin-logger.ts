import { Ansis, blue, cyanBright, green, hex, red } from "ansis";
import { inspect } from "node:util";
import { performance } from "node:perf_hooks";
import { FlatObject, isProduction } from "./common-helpers";
import {
  AbstractLogger,
  isHidden,
  makeNumberFormat,
  Severity,
} from "./logger-helpers";

interface Context extends FlatObject {
  requestId?: string;
}

export interface BuiltinLoggerConfig {
  /**
   * @desc The minimal severity to log or "silent" to disable logging
   * @example "debug" also enables pretty output for inspected entities
   * */
  level: "silent" | "warn" | "info" | "debug";
  /** @desc Enables colors on printed severity and inspected entities */
  color: boolean;
  /**
   * @desc Control how deeply entities should be inspected
   * @example null
   * @example Infinity
   * */
  depth: number | null;
  /**
   * @desc Context: the metadata applicable for each logged entry, used by .child() method
   * @see childLoggerProvider
   * */
  ctx: Context;
}

interface ProfilerOptions {
  message: string;
  /** @default "debug" */
  severity?: Severity | ((ms: number) => Severity);
  /** @default formatDuration - adaptive units and limited fraction */
  formatter?: (ms: number) => string | number;
}

/** @desc Built-in console logger with optional colorful inspections */
export class BuiltinLogger implements AbstractLogger {
  protected readonly picoFormat = makeNumberFormat("nanosecond", 3);
  protected readonly nanoFormat = makeNumberFormat("nanosecond");
  protected readonly microFormat = makeNumberFormat("microsecond");
  protected readonly milliFormat = makeNumberFormat("millisecond");
  protected readonly secondsFormat = makeNumberFormat("second", 2);
  protected readonly minutesFormat = makeNumberFormat("minute", 2);
  protected readonly config: BuiltinLoggerConfig;
  protected readonly styles: Record<Severity, Ansis> = {
    debug: blue,
    info: green,
    warn: hex("#FFA500"),
    error: red,
  };

  /** @example new BuiltinLogger({ level: "debug", color: true, depth: 4 }) */
  public constructor(config: Partial<BuiltinLoggerConfig> = {}) {
    const {
      color = new Ansis().isSupported(),
      level = isProduction() ? "warn" : "debug",
      depth = 2,
      ctx = {},
    } = config;
    this.config = { color, level, depth, ctx };
  }

  protected prettyPrint(subject: unknown) {
    const { depth, color: colors, level } = this.config;
    return inspect(subject, {
      depth,
      colors,
      breakLength: level === "debug" ? 80 : Infinity,
      compact: level === "debug" ? 3 : true,
    });
  }

  protected print(method: Severity, message: string, meta?: unknown) {
    const {
      level,
      ctx: { requestId, ...ctx },
      color: hasColor,
    } = this.config;
    if (level === "silent" || isHidden(method, level)) return;
    const output: string[] = [new Date().toISOString()];
    if (requestId) output.push(hasColor ? cyanBright(requestId) : requestId);
    output.push(
      hasColor ? `${this.styles[method](method)}:` : `${method}:`,
      message,
    );
    if (meta !== undefined) output.push(this.prettyPrint(meta));
    if (Object.keys(ctx).length > 0) output.push(this.prettyPrint(ctx));
    console.log(output.join(" "));
  }

  public debug(message: string, meta?: unknown) {
    this.print("debug", message, meta);
  }

  public info(message: string, meta?: unknown) {
    this.print("info", message, meta);
  }

  public warn(message: string, meta?: unknown) {
    this.print("warn", message, meta);
  }

  public error(message: string, meta?: unknown) {
    this.print("error", message, meta);
  }

  public child(ctx: Context) {
    return new BuiltinLogger({ ...this.config, ctx });
  }

  protected formatDuration = (ms: number) => {
    if (ms < 1e-6) return this.picoFormat.format(ms / 1e-6);
    if (ms < 1e-3) return this.nanoFormat.format(ms / 1e-6);
    if (ms < 1) return this.microFormat.format(ms / 1e-3);
    if (ms < 1e3) return this.milliFormat.format(ms);
    if (ms < 6e4) return this.secondsFormat.format(ms / 1e3);
    return this.minutesFormat.format(ms / 6e4);
  };

  /** @desc Measures the duration until you invoke the returned callback */
  public profile(message: string): () => void;
  public profile(options: ProfilerOptions): () => void;
  public profile(subject: string | ProfilerOptions) {
    const start = performance.now();
    return () => {
      const duration = performance.now() - start;
      const {
        message,
        severity = "debug",
        formatter = this.formatDuration.bind(this),
      } = typeof subject === "object" ? subject : { message: subject };
      this.print(
        typeof severity === "function" ? severity(duration) : severity,
        message,
        formatter(duration),
      );
    };
  }
}
