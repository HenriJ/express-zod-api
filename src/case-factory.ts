import { z } from "zod";
import { Case, Handler } from "./case";

export interface CaseDefinifion<
  IN extends z.ZodTuple,
  OUT extends z.ZodTuple | undefined,
> {
  input: IN;
  output?: OUT;
  handler: Handler<z.output<IN>, OUT extends z.ZodTuple ? z.input<OUT> : void>;
}

export class CaseFactory {
  public build<IN extends z.ZodTuple, OUT extends z.ZodTuple | undefined>(
    def: CaseDefinifion<IN, OUT>,
  ) {
    return new Case(def);
  }
}
