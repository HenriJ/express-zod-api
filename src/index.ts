export { ConfigType } from './config-type';
export { AbstractEndpoint, Method } from './endpoint';
export { EndpointsFactory } from './endpoints-factory';
export { ObjectSchema, Unshape } from './helpers';
export { createLogger } from './logger';
export { createMiddleware } from './middleware';
export { ResultHandler } from './result-handler';
export { Routing } from './routing';
export { createServer } from './server';

import * as z from 'zod';
import * as createHttpError from 'http-errors';

export { createHttpError, z };
