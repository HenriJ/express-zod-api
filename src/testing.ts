import { Request, Response } from "express";
import http from "node:http";
import { CommonConfig } from "./config-type";
import { AbstractEndpoint } from "./endpoint";
import { AbstractLogger } from "./logger";
import { mimeJson } from "./mime";

type MockFunction = <S>(implementation?: (...args: any[]) => any) => S; // kept "any" for easier compatibility

export const makeRequestMock = <
  REQ extends Record<string, any>,
  FN extends MockFunction,
>({
  fnMethod,
  requestProps,
}: {
  fnMethod: FN;
  requestProps?: REQ;
}) =>
  ({
    method: "GET",
    header: fnMethod(() => mimeJson),
    ...requestProps,
  }) as { method: string } & Record<"header", ReturnType<FN>> & REQ;

export const makeResponseMock = <
  RES extends Record<string, any>,
  FN extends MockFunction,
>({
  fnMethod,
  responseProps,
}: {
  fnMethod: FN;
  responseProps?: RES;
}) => {
  const responseMock = {
    writableEnded: false,
    statusCode: 200,
    statusMessage: http.STATUS_CODES[200],
    set: fnMethod(() => responseMock),
    setHeader: fnMethod(() => responseMock),
    header: fnMethod(() => responseMock),
    status: fnMethod((code) => {
      responseMock.statusCode = code;
      responseMock.statusMessage = http.STATUS_CODES[code]!;
      return responseMock;
    }),
    json: fnMethod(() => responseMock),
    send: fnMethod(() => responseMock),
    end: fnMethod(() => {
      responseMock.writableEnded = true;
      return responseMock;
    }),
    ...responseProps,
  } as {
    writableEnded: boolean;
    statusCode: number;
    statusMessage: string;
  } & Record<
    "set" | "setHeader" | "header" | "status" | "json" | "send" | "end",
    ReturnType<FN>
  > &
    RES;
  return responseMock;
};

interface TestEndpointProps<REQ, RES, LOG, FN> {
  /** @desc The endpoint to test */
  endpoint: AbstractEndpoint;
  /**
   * @desc Additional properties to set on Request mock
   * @default { method: "GET", header: () => "application/json" }
   * */
  requestProps?: REQ;
  /**
   * @desc Additional properties to set on Response mock
   * @default { writableEnded, statusCode, statusMessage, set, setHeader, header, status, json, send, end }
   * */
  responseProps?: RES;
  /**
   * @desc Additional properties to set on config mock
   * @default { cors: false, logger }
   * */
  configProps?: Partial<CommonConfig>;
  /**
   * @desc Additional properties to set on logger mock
   * @default { info, warn, error, debug }
   * */
  loggerProps?: LOG;
  /**
   * @example jest.fn
   * @example vi.fn
   * */
  fnMethod: FN;
}

/**
 * @desc You need to install either jest (with @types/jest) or vitest in order to use this method
 * @requires jest
 * @requires vitest
 * */
export const testEndpoint = async <
  FN extends MockFunction,
  LOG extends Record<string, any>,
  REQ extends Record<string, any>,
  RES extends Record<string, any>,
>({
  endpoint,
  requestProps,
  responseProps,
  configProps,
  loggerProps,
  fnMethod,
}: TestEndpointProps<REQ, RES, LOG, FN>) => {
  const requestMock = makeRequestMock({ fnMethod: fnMethod, requestProps });
  const responseMock = makeResponseMock({ fnMethod: fnMethod, responseProps });
  const loggerMock = {
    info: fnMethod(),
    warn: fnMethod(),
    error: fnMethod(),
    debug: fnMethod(),
    ...loggerProps,
  } as Record<keyof AbstractLogger, ReturnType<FN>> & LOG;
  const configMock = {
    cors: false,
    logger: loggerMock,
    ...configProps,
  };
  await endpoint.execute({
    request: requestMock as unknown as Request,
    response: responseMock as unknown as Response,
    config: configMock as CommonConfig,
    logger: loggerMock as AbstractLogger,
  });
  return { requestMock, responseMock, loggerMock };
};
