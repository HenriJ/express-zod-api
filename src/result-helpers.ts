import { Request } from "express";
import { isHttpError } from "http-errors";
import assert from "node:assert/strict";
import { z } from "zod";
import { NormalizedResponse, ResponseVariant } from "./api-response";
import { FlatObject } from "./common-helpers";
import { InputValidationError, ResultHandlerError } from "./errors";
import { ActualLogger } from "./logger-helpers";
import type { LazyResult, Result } from "./result-handler";

export type ResultSchema<R extends Result> =
  R extends Result<infer S> ? S : never;

/** @throws ResultHandlerError when Result is an empty array */
export const normalize = <A extends unknown[]>(
  subject: Result | LazyResult<Result, A>,
  features: {
    variant: ResponseVariant;
    arguments: A;
    statusCodes: [number, ...number[]];
    mimeTypes: [string, ...string[]];
  },
): NormalizedResponse[] => {
  if (typeof subject === "function") {
    return normalize(subject(...features.arguments), features);
  }
  if (subject instanceof z.ZodType) {
    return [{ ...features, schema: subject }];
  }
  if (Array.isArray(subject)) {
    assert(
      subject.length,
      new ResultHandlerError(
        new Error(`At least one ${features.variant} response schema required.`),
      ),
    );
  }
  return (Array.isArray(subject) ? subject : [subject]).map(
    ({ schema, statusCodes, statusCode, mimeTypes, mimeType }) => ({
      schema,
      statusCodes: statusCode
        ? [statusCode]
        : statusCodes || features.statusCodes,
      mimeTypes: mimeType ? [mimeType] : mimeTypes || features.mimeTypes,
    }),
  );
};

export const isServerSideIssue = (statusCode: number) =>
  ~~(statusCode / 100) === 5; // 5XX

export const logServerError = ({
  logger,
  request,
  input,
  error,
  statusCode,
}: {
  logger: ActualLogger;
  request: Request;
  input: FlatObject | null;
  error: Error;
  statusCode: number;
}) =>
  isServerSideIssue(statusCode) &&
  logger.error("Server side error", {
    error,
    url: request.url,
    payload: input,
  });

export const getStatusCodeFromError = (error: Error): number => {
  if (isHttpError(error)) return error.statusCode;
  return error instanceof InputValidationError ? 400 : 500;
};
