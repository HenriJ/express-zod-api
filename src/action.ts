import { init, last } from "ramda";
import type { Socket } from "socket.io";
import { z } from "zod";
import { AckActionDef, EmissionMap, SimpleActionDef } from "./actions-factory";
import { InputValidationError, OutputValidationError } from "./errors";
import { AbstractLogger } from "./logger";

export type Handler<IN, OUT, E extends EmissionMap> = (params: {
  input: IN;
  logger: AbstractLogger;
  emit: <K extends keyof E>(
    evt: K,
    ...args: z.input<E[K]["schema"]>
  ) => Promise<
    z.output<
      E[K]["ack"] extends z.AnyZodTuple ? E[K]["ack"] : z.ZodLiteral<true>
    >
  >;
  isConnected: () => boolean;
}) => Promise<OUT>;

export abstract class AbstractAction {
  public abstract execute(params: {
    event: string;
    params: unknown[];
    logger: AbstractLogger;
    socket: Socket;
  }): Promise<void>;
}

export class Action<
  IN extends z.AnyZodTuple,
  OUT extends z.AnyZodTuple,
  E extends EmissionMap,
> extends AbstractAction {
  readonly #inputSchema: IN;
  readonly #outputSchema: OUT | undefined;
  readonly #handler: Handler<z.output<IN>, z.input<OUT> | void, E>;
  readonly #emission: E;

  public constructor(
    action: AckActionDef<IN, OUT, E> | SimpleActionDef<IN, E>,
    emission: E,
  ) {
    super();
    this.#inputSchema = action.input;
    this.#outputSchema = "output" in action ? action.output : undefined;
    this.#handler = action.handler;
    this.#emission = emission;
  }

  async #emit<K extends keyof E>({
    event,
    args,
    logger,
    socket,
  }: {
    event: K;
    args: z.input<E[K]["schema"]>;
    logger: AbstractLogger;
    socket: Socket;
  }): Promise<
    z.output<
      E[K]["ack"] extends z.AnyZodTuple ? E[K]["ack"] : z.ZodLiteral<true>
    >
  > {
    const { schema, ack: ackSchema } = this.#emission[event];
    const emitValidation = schema.safeParse(args);
    if (!emitValidation.success) {
      return logger.error(
        `${String(event)} emission validation error`,
        new OutputValidationError(emitValidation.error),
      );
    }
    logger.debug(`Emitting ${String(event)}`, emitValidation.data);
    if (!ackSchema) {
      return socket.emit(String(event), ...emitValidation.data) || true;
    }
    try {
      const ack = await socket
        .timeout(2000) // @todo configurable?
        .emitWithAck(String(event), ...emitValidation.data);
      return ackSchema.parse(ack);
    } catch (error) {
      if (error instanceof z.ZodError) {
        throw new InputValidationError(error);
      }
      throw error;
    }
  }

  public override async execute({
    event,
    params,
    logger,
    socket,
  }: {
    event: string;
    params: unknown[];
    logger: AbstractLogger;
    socket: Socket;
  }): Promise<void> {
    const payload = this.#outputSchema ? init(params) : params;
    const inputValidation = this.#inputSchema.safeParse(payload);
    if (!inputValidation.success) {
      return logger.error(
        `${event} payload validation error`,
        new InputValidationError(inputValidation.error),
      );
    }
    logger.debug(
      `parsed input (${this.#outputSchema ? "excl." : "no"} ack)`,
      inputValidation.data,
    );
    const ackValidation = this.#outputSchema
      ? z.function(this.#outputSchema, z.void()).safeParse(last(params))
      : undefined;
    if (ackValidation && !ackValidation.success) {
      return logger.error(
        `${event} acknowledgement validation error`,
        new InputValidationError(ackValidation.error),
      );
    }
    const ack = ackValidation?.data;
    const output = await this.#handler({
      input: inputValidation.data,
      logger,
      emit: (evt, ...args) => this.#emit({ event: evt, args, logger, socket }),
      isConnected: () => socket.connected,
    });
    if (!this.#outputSchema) {
      return; // no ack
    }
    const outputValidation = this.#outputSchema.safeParse(output);
    if (!outputValidation.success) {
      return logger.error(
        `${event} output validation error`,
        new OutputValidationError(outputValidation.error),
      );
    }
    logger.debug("parsed output", outputValidation.data);
    if (ack) {
      ack(...outputValidation.data);
    }
  }
}
