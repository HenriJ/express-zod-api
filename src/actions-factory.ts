import { z } from "zod";
import { Action, Handler } from "./action";
import { CommonConfig } from "./config-type";
import { EmissionMap } from "./emission";

export interface SimpleActionDef<
  IN extends z.AnyZodTuple,
  E extends EmissionMap,
> {
  input: IN;
  handler: Handler<z.output<IN>, void, E>;
}

export interface AckActionDef<
  IN extends z.AnyZodTuple,
  OUT extends z.AnyZodTuple,
  E extends EmissionMap,
> {
  input: IN;
  output: OUT;
  handler: Handler<z.output<IN>, z.input<OUT>, E>;
}

export class ActionsFactory<E extends EmissionMap> {
  constructor(protected config: CommonConfig<string, E>) {}

  public build<IN extends z.AnyZodTuple, OUT extends z.AnyZodTuple>(
    def: SimpleActionDef<IN, E> | AckActionDef<IN, OUT, E>,
  ): Action<IN, OUT> {
    return new Action(def);
  }
}
