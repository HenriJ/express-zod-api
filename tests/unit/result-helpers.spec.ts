import { range } from "ramda";
import { isServerSideIssue, logServerError } from "../../src/result-helpers";
import { makeLoggerMock, makeRequestMock } from "../../src/testing";

describe("Result helpers", () => {
  describe("isServerSideIssue()", () => {
    test.each(range(100, 599))(
      "should be true when %i >= 500",
      (statusCode) => {
        expect(isServerSideIssue(statusCode)).toBe(statusCode >= 500);
      },
    );
  });

  describe("logServerError()", () => {
    test("should log server side error", () => {
      const error = new Error("test");
      const logger = makeLoggerMock();
      const request = makeRequestMock({ url: "https://example.com" });
      logServerError({
        error,
        logger,
        request,
        statusCode: 501,
        input: { test: 123 },
      });
      expect(logger._getLogs().error).toEqual([
        [
          "Server side error",
          { error, payload: { test: 123 }, url: "https://example.com" },
        ],
      ]);
    });
  });
});
