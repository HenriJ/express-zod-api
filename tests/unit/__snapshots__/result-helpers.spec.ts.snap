// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Result helpers > ensureHttpError() > should handle Error: basic 1`] = `
InternalServerError({
  "cause": Error({
    "message": "basic",
  }),
  "message": "basic",
})
`;

exports[`Result helpers > ensureHttpError() > should handle NotFoundError: Not really found 1`] = `
NotFoundError({
  "message": "Not really found",
})
`;

exports[`Result helpers > ensureHttpError() > should handle ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
] { issues: [ { code: 'invalid_type', expected: 'string', received: 'number', path: [], message: 'Expected string, received number' } ], addIssue: [Function], addIssues: [Function] } 1`] = `
InternalServerError({
  "cause": ZodError({
    "message": "[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]",
  }),
  "message": "Expected string, received number",
})
`;
